multicluster/controllers/multicluster/leader/member_cleanup_controller.go

```diff
@@ -0,0 +1,218 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package leader
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	mcv1alpha1 "antrea.io/antrea/multicluster/apis/multicluster/v1alpha1"
+)
+
+const indexKey = "spec.clusterID"
+
+var (
+	getResourceExportsByClusterIDFunc = getResourceExportsByClusterID
+
+	// cleanUpRetry is the retry when the cleanUpStaleResourceExports
+	// failed to clean up all stale ResourceExports.
+	cleanUpRetry = wait.Backoff{
+		Steps:    5,
+		Duration: 500 * time.Millisecond,
+		Factor:   1.0,
+		Jitter:   1,
+	}
+
+	ignoreUpdateEventPredicate = predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+	}
+)
+
+// MemberResourcesCleanupController will remove all ResourceExports belong to a member
+// cluster when the corresponding MemberClusterAnnounce CR is deleted. It will also try
+// to clean up all stale ResourceExports during start.
+type MemberResourcesCleanupController struct {
+	client.Client
+	Scheme    *runtime.Scheme
+	namespace string
+}
+
+func NewMemberResourcesCleanupController(Client client.Client,
+	Scheme *runtime.Scheme,
+	namespace string) *MemberResourcesCleanupController {
+	controller := MemberResourcesCleanupController{
+		Client:    Client,
+		Scheme:    Scheme,
+		namespace: namespace,
+	}
+	return &controller
+}
+
+func (c *MemberResourcesCleanupController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	memberAnnounce := &mcv1alpha1.MemberClusterAnnounce{}
+	err := c.Get(ctx, req.NamespacedName, memberAnnounce)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, err
+	}
+
+	if err == nil && memberAnnounce.DeletionTimestamp.IsZero() {
+		// Try to clean up stale ResourceExports whenever there is a MemberClusterAnnounce
+		// creation event, to ensure that stale resources are removed in time so that
+		// there is no stale data for ResourceExport reconciler and being imported
+		// into member clusters incorrectly.
+		err := c.cleanUpStaleResourceExports(ctx)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
```

Comment:
Hi @jianjuns beside the code re-org, two changes are involved to ensure resource cleanup.
Above changes are making sure there is no stale ResourceExports when a new member cluster joined.

Reply:
Do we have any synchronization between cleanup and ResourceExport controllers, so ResourceExport controller wont run before cleanup?

And cleanUpStaleResourceExports() can fail right? In that case, ResourceExport may still see stale Exports? Then do we really need the special handling here to do cleanup before ResourceExport controller?

Reply:
I doubt it would help for avoiding stale resources by setting up synchronization between cleanup and ResourceExport controllers. Eg: when member cluster A is deleted, another member cluster B may have some Exports updates. If we only allow ResourceExport controllers to do reconcile after cleanup is done. There are still out-of-date data for member cluster B if there is any failure happen during cleanup.

I feel it's hard to avoid the stale Exports completely, eg: leader API is not stable and member fail to update the exports in time, or member controller is in a crashed status for a while but less than 24 hrs etc. 
The purpose is to ensure that we can clean up stale ResourceExports in every chance. I think it would be helpful to do cleanup when a new member cluster is joining, it usually means no exports yet but a chance to do another cleanup. It help to ensure that the status of a ClusterSet can be right eventually.

Reply:
If there is no guarantee no stale resource before ResourceExport reconciliation, I do not feel very useful to do cleanup before a new member. We should just do cleanup when a member leaves, and rely on stale controller (?) to delete stale resources for other cases.

Reply:
OK, removed this part.